ui: embed UI assets with go:embed directive

koorosh opened the pull request:
Purpose of this change is to simplify build process for both
`make` and `bazel` builds.
Initially, ui assets were linked into go binaries by go-bindata
and required additional steps to format and
append extra data in auto-generated file that does not
work well with bazel builds.
To avoid these complications during build, new go:embed directive
is used instead.

Following gotchas required to make it work:
- all generated JS assets are placed into `assets` subdirectory
which is then linked as embedded dir.
- go:embed doesn't allow to point on empty or non-existing dir
 so `assets` dir includes dummy `index.html` file (the same as
 template for short builds without UI and .gitkeep file for
 convention to make it clear that this dir should be checked in
 repository.

Release note: none

Release justification: non-production code changes

koorosh committed:
ui: embed UI assets with go:embed directive

Purpose of this change is to simplify build process for both
`make` and `bazel` builds.
Initially, ui assets were linked into go binaries by go-bindata
and required additional steps to format and
append extra data in auto-generated file that does not
work well with bazel builds.
To avoid these complications during build, new go:embed directive
is used instead.

Following gotchas required to make it work:
- all generated JS assets are placed into `assets` subdirectory
which is then linked as embedded dir.
- go:embed doesn't allow to point on empty or non-existing dir
 so `assets` dir includes dummy `index.html` file (the same as
 template for short builds without UI and .gitkeep file for
 convention to make it clear that this dir should be checked in
 repository.

Makefile recipes for building ui binaries are adjusted to work
with these changes.
(!) It is important to mention one downside - now UI recipe is
a phony target and will be executed every time for `cockroach`
binaries. It should be reconsidered how to avoid this behavior.

Release note: none

Release justification: non-production code changes

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69135)
<!-- Reviewable:end -->

Santamaura reviewed (approved):
LGIT

> Reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69135#-:-MhV6Sth0bCYFSc4okgv:bb6ywdn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nathanstilwell and @rickystewart)



<!-- Sent from Reviewable.io -->

koorosh committed:
ui: remove go-bindata-assetfs from go modules

Clean up go modules and remove unused module

Release note: None

Release justification: non-production code changes

rickystewart commented on Makefile:
```diff
@@ -1403,18 +1402,16 @@ ui-test-watch: $(UI_CCL_DLLS) $(UI_CCL_MANIFESTS)
+pkg/ui/assets.oss.installed: $(UI_OSS_DLLS) $(UI_OSS_MANIFESTS) $(UI_JS_OSS)
+pkg/ui/assets.%.installed: pkg/ui/webpack.app.js $(shell find pkg/ui/src pkg/ui/styl -type f) | bin/.bootstrap
+	find pkg/ui/dist$*/assets -mindepth 1 -not -name .gitkeep -not -name index.html -delete
+	for dll in $(shell find pkg/ui/dist/*.dll.js -type f); do \
```
`$(shell find pkg/ui/dist -name '*.dll.js' -type f)`

rickystewart commented on pkg/ui/distoss/distoss.go:
```diff
@@ -11,3 +11,17 @@
+	"github.com/cockroachdb/cockroach/pkg/ui"
+)
+
+//go:embed assets
```
Just making sure, this technique can work with Bazel? I've never seen it before.

koorosh replied:
Go Source rule includes `embedsrcs` field which allows to deal with `go:embed` directive:
https://github.com/bazelbuild/rules_go/blob/master/go/providers.rst#gosource

rickystewart commented on pkg/ui/.gitignore:
```diff
@@ -8,10 +8,14 @@ yarn.opt.installed
 # Generated intermediates
 .cache-loader
 dist*/**
+!dist*/.gitkeep
```
What purpose does the `.gitkeep` file serve? From what I can tell, since we also have the `index.html` file in the same directory, it doesn't need to be there.

koorosh replied:
Correct, it has to be removed, thanks!

rickystewart reviewed (changes requested):
Mostly looks good, just a couple quick questions/clarifications/nits.

koorosh commented:
bors r+
